src/fs.rs

```diff
@@ -64,14 +67,40 @@ pub fn make_temp_dir(
   }
 }
 
-pub fn mkdir(path: &Path) -> std::io::Result<()> {
+pub fn mkdir(path: &Path, perm: u32) -> std::io::Result<()> {
   debug!("mkdir -p {}", path.display());
-  assert!(path.has_root(), "non-has_root not yet implemented");
-  std::fs::create_dir_all(path).or_else(|err| {
-    if err.kind() == std::io::ErrorKind::AlreadyExists {
-      Ok(())
-    } else {
-      Err(err)
+  let mut builder = DirBuilder::new();
+  builder.recursive(true);
+  set_dir_permission(&mut builder, perm);
+  builder.create(path).or_else(|err| match err.kind() {
+    std::io::ErrorKind::AlreadyExists => Ok(()),
+    _ => Err(err),
+  })
+}
+
+#[cfg(any(unix))]
+fn set_dir_permission(builder: &mut DirBuilder, perm: u32) {
+  debug!("set dir perm to {}", perm);
+  builder.mode(perm);
+}
+
+#[cfg(not(any(unix)))]
+fn set_dir_permission(_builder: &mut DirBuilder, _perm: u32) {
+  // NOOP on windows
+}
+
+pub fn rmdir(path: &Path, recursive: bool) -> std::io::Result<()> {
```

Comment:
Remove this function, everything should be handled in `handle_remove` inside src/handlers.rs

Reply:
Done.